text/0000-export.md

```diff
@@ -0,0 +1,401 @@
+The hash algorithm is still to be determined.
+
+(See also the "alternatives" section below.)
+
+### Type Information
+
+As mentioned above, the hash in a symbol name needs to cover _all_ relevant type information.
+However, exactly which information is and isn't relevant for safety is a complicated question.
+
+#### Types with Public Fields
+
+For a simple user defined type where all fields are both public, like the `Point` struct below,
+the relevant parts are the size, alignment, and recursively all field information.
+
+```rust
+#[export]
+#[repr(C)]
+pub struct Point {
+    pub x: f32,
+    pub y: f32,
+    pub name: &str,
+}
+```
+
+The `#[export]` attribute is the user's commitment to keep the type stable, but without `unsafe`,
+any mistakes should _not_ result in unsoundness.
+Accidentally changing the struct to swap the `x` and `name` fields should result in a different hash,
+such that the `f32` won't get interpreted as a `&str`, for example.
+
+Note that, technically, the names of the type and the fields are not relevant, at least _not for memory safety_.
+Swapping the `x` and `y` fields result in surprises and bugs and shouldn't be done,
+but it won't result in undefined behaviour, since any Rust code can swap the fields without using `unsafe`.
+
+However, for public fields, the field names are already part of the stable API, so we include them in the hash as well.
+
+It is an error to use a plain `#[export]` attribute on a type with out stable `#[repr(…)]`,
+if it has any private fields,
+or if any of the fields are not of an `#[export]`ed or builtin type.
+
+#### Types with Private Fields
+
+For types where not all fields are public, the situation is much more complicated.
+
+Private fields usually come with certain *invariants*, and come with `unsafe` code that makes assumptions about them.
+For example, the private fields of a `Vec` are assumed to represent a valid "owned" pointer to an allocation together with its capacity and initialized size.
+
+If it would be possible to define a identically named type with the same fields but different (or no) invariants/assumptions,
+or just change the invariants in an existing library,
+it'd be possible to cause undefined behavior by loading the "wrong" dynamic library.
+
+Therefore, we can't allow a regular `#[export]` attribute on a type with private fields,
+since we have no way of automatically determining the invariants / unsafe assumptions about private fields.
+
+Instead, for these types, we must require the user to *unsafely* commit to
+ABI stability if they want to make the type available to exported functions.
+
+Using `#[export(unsafe_stable_abi = «hash»)]`, one can make the (unsafe) promise
+that the type will remain ABI compatible as long as the provided hash remains the same.
+The hash must have been randomly generated to ensure uniqueness (which is part of the unsafe promise).
```

Comment:
As a cryptographic engineer, calling this a `hash` is confusing to me; as specified, it isn't a hash of anything, but instead is sampled randomness. I'm guessing from the naming that the intention is for this user-provided random value to take the place of the hash that a fully-public type would have in its mangled name?

Perhaps `commitment` would be a more clarifying name: for public types, the API's commitment is made via a hash of the public API, and for private-containing types, the commitment is made directly and unsafely by the user.

Reply:
> I'm guessing from the naming that the intention is for this user-provided random value to take the place of the hash that a fully-public type would have in its mangled name?

Yes.

It represents the hash of all relevant memory safety information, which for types with private fields includes all safety invariants. Because those are not expressed directly in the type system and we cannot hash someone's mental model, this just uses a random number that the user promises to change when the inveriants change. :)

Happy to call it something else, but the name of this placeholder isn't really an important part of the proposal.

Reply:
I don't see why this needs to be a hash or big random number at all, and it is harmful for readability and will cause all sorts of auditability problems (is this truly random?) Why can't this just be an extra salt on top of all the usual information that goes into the ABI hash for the type. Then it is effectively just "ABI_version = 4" or what have you, you should bump it when something changes and this is easy for users to understand.